test/distributed/pipelining/test_backward.py

```diff
@@ -117,22 +135,33 @@ def test_stage_backward_weight(self):
         )
 
         # backward of loss with respect to weights
-        dweights = stage_backward_weight(mod.parameters(), param_groups)
+        stage_backward_weight(mod.parameters(), param_groups, retain_graph=True)
 
         # Run reference
         ref_out = ref_mod(ref_x)
         ref_loss = loss_fn(ref_out, ref_target)
         ref_loss.backward()
 
         # Every rank checks gradients
-        for name, p in mod.named_parameters():
+        for i, (name, p) in enumerate(mod.named_parameters()):
             ref_p = ref_mod.get_parameter(name)
             try:
                 torch.testing.assert_close(p.grad, ref_p.grad)
             except AssertionError:
                 print(f"Gradient test failed for {name}: {p.grad} vs {ref_p.grad}")
                 raise
 
+        # Monitor memory after backward_weight
+        for _ in range(5):
+            gc.collect()
+            before_gc_count = len(gc.get_objects())
+            out = mod(x)
+            loss = loss_fn(out, target)
+            stage_backward_weight(mod.parameters(), param_groups, retain_graph=True)
+            gc.collect()
+            after_gc_count = len(gc.get_objects())
+            self.assertLessEqual(after_gc_count, before_gc_count)
```

Comment:
are you monitoring python memory? We probably want to monitor gpu memory and have some tighter bound about the expected size that gets freed